src/sp1_soldering.rs

```diff
@@ -0,0 +1,395 @@
+    fn build_expected_public_params(
+        raw_instances: &[Vec<(u128, u128)>],
+        nonce: u128,
+    ) -> SolderedLabels {
+        let (base, rest) = raw_instances
+            .split_first()
+            .expect("at least one instance required");
+
+        let mut base_commitment = Vec::with_capacity(base.len());
+        let mut base_nonce_commitment = Vec::with_capacity(base.len());
+
+        for &(label0, label1) in base {
+            base_commitment.push((hash_label(label0), hash_label(label1)));
+            base_nonce_commitment.push((hash_label(label0 ^ nonce), hash_label(label1 ^ nonce)));
+        }
+
+        let mut commitments = Vec::with_capacity(rest.len());
+        let mut deltas = Vec::with_capacity(rest.len());
+
+        for instance in rest {
+            let mut instance_commitment = Vec::with_capacity(instance.len());
+            let mut instance_deltas = Vec::with_capacity(instance.len());
+
+            for ((base0, base1), (label0, label1)) in base.iter().zip(instance.iter()) {
+                instance_commitment.push((hash_label(*label0), hash_label(*label1)));
+                instance_deltas.push((base0 ^ label0, base1 ^ label1));
+            }
+
+            commitments.push(instance_commitment);
+            deltas.push(instance_deltas);
+        }
+
+        SolderedLabels {
+            deltas,
+            base_commitment,
+            base_nonce_commitment,
+            nonce,
+            commitments,
+        }
+    }
+
+    #[test]
+    #[ignore]
+    fn test_core_public_params_roundtrip() {
+        use rand::{Rng, SeedableRng, rngs::StdRng};
+
+        let wires = 1;
+        let instances = 1;
+
+        let mut rng = StdRng::seed_from_u64(42);
+        let nonce: u128 = rng.r#gen();
+        let delta: u128 = rng.r#gen::<u128>() | 1;
+
+        let mut raw_instances = Vec::with_capacity(1 + instances);
+
+        for _ in 0..=instances {
+            let mut raw_labels = Vec::with_capacity(wires);
+
+            for _ in 0..wires {
+                let label0: u128 = rng.r#gen();
+                let label1: u128 = label0 ^ delta;
+                raw_labels.push((label0, label1));
+            }
+
+            raw_instances.push(raw_labels);
+        }
+
+        let input = types::WiresInput {
+            instances_wires: raw_instances.clone(),
+            nonce,
+        };
+
+        let input_bytes =
+            serialize_wires_input(&input).expect("failed to serialize wires input for core test");
+
+        let mut stdin = SP1Stdin::new();
+        stdin.write(&input_bytes.as_slice());
+
+        let prover = SP1Prover::<CpuProverComponents>::new();
+        let (_pk, pk_device, program, _vk) = prover.setup(elf());
+        let opts = SP1ProverOpts::default();
+        let context = SP1ContextBuilder::default().build();
+
+        let core_proof = prover
+            .prove_core(&pk_device, program, &stdin, opts, context)
+            .expect("core proving failed");
```

Comment:
The test is generating a real SNARK proof and will take very long time. Instead, we can either

- Use `prover.execute` which executes the circuit without really genertaing a proof
- Use SP1's mock mode `SP1_PROVER=mock` such that `prover.prove_core` will only execute circuit

SP1 guarantees that if the circuit can be executed successfully, then one can generate proof with the same input.